plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php

```diff
@@ -800,39 +797,24 @@ private function set_array_meta( $key, $value, $object ) {
 			$meta_key = self::ADDITIONAL_FIELDS_KEY;
 		}
 
-		if ( $object instanceof \WC_Customer ) {
-			if ( ! $object->get_id() ) {
-				$meta_data = wc()->session->get( $meta_key, array() );
-			} else {
-				$meta_data = get_user_meta( $object->get_id(), $meta_key, true );
-			}
-		} elseif ( $object instanceof \WC_Order ) {
-			$meta_data = $object->get_meta( $meta_key, true );
-		}
+		$meta_data = $object->get_meta( $meta_key, true );
 
 		if ( ! is_array( $meta_data ) ) {
 			$meta_data = array();
 		}
 
 		$meta_data[ $key ] = $value;
-		if ( $object instanceof \WC_Customer ) {
-			if ( ! $object->get_id() ) {
-				wc()->session->set( $meta_key, $meta_data );
-			} else {
-				update_user_meta( $object->get_id(), $meta_key, $meta_data );
-			}
-		} elseif ( $object instanceof \WC_Order ) {
-			$object->update_meta_data( $meta_key, $meta_data );
-		}
 
+		// Replacing all meta using `add_meta_data`. For some reason `update_meta_data` causes duplicate keys.
+		$object->add_meta_data( $meta_key, $meta_data, true );
```

Comment:
The true here should handle existing meta.

Reply:
This ensures the meta overwrites existing meta to prevent duplicates. Otherwise you get duplicates of our top level keys.